### PR TITLE
[AIRFLOW-4401] Use managers for Queue synchronization

### PR DESCRIPTION
**Pull request by potiuk**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4401
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

It is a known problem https://bugs.python.org/issue23582 that
multiprocessing.Queue empty() method is not reliable - sometimes it might
return True even if another process already put something in the queue.

This resulted in some of the tasks not picked up when sync() methods
were called (in AirflowKubernetesScheduler, LocalExecutor,
DagFileProcessor). This was less of a problem if the method was called in sync()
- as the remaining jobs/files could be processed in next pass but it was a problem
in tests and when graceful shutdown was executed (some tasks could be still
unprocessed while the shutdown occured).

We switched to Managers() managed queues to handle that - the queue in this case
is run in a separate subprocess and each process using it uses a proxy to access
this shared queue.

All the cases impacted follow the same pattern now:

while not queue.empty():
   res = queue.get()
   ....

This loop runs always in single (main) process so it is safe to run it this way -
there is no risk that some other process will retrieve the data from the queue in
between empty() and get().

In all these cases overhead for inter-processing locking is negligible
comparing to the action executed (Parsing DAG, executing job)
so it appears it should be safe to merge the change.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Tests were there (but flaky)

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: tests/test_jobs.py [hunk 1]
```diff
@@ -206,7 +206,7 @@ def test_trigger_controller_dag(self):
 
         scheduler = SchedulerJob()
         queue = Mock()
-        scheduler._process_task_instances(target_dag, queue=queue)
+        scheduler._process_task_instances(target_dag, task_instances_list=queue)
```
Comment (potiuk) [inline tests/test_jobs.py]: I will rename all the queues later when we discuss which solution to choose.

File: airflow/executors/base_executor.py [hunk 2]
```diff
@@ -147,7 +147,6 @@ def heartbeat(self):
         self.sync()
 
     def change_state(self, key, state):
-        self.log.debug("Changing state: %s", key)
```
Comment (ashb) [inline airflow/executors/base_executor.py]: Unrelated change here (I usually try to avoid these.)

**Review by ashb — APPROVED** [review]
Looks good though

File: airflow/contrib/executors/kubernetes_executor.py [hunk 3]
```diff
@@ -373,7 +372,8 @@ def __init__(self, kube_config, task_queue, result_queue, kube_client, worker_uu
         self.kube_client = kube_client
         self.launcher = PodLauncher(kube_client=self.kube_client)
         self.worker_configuration = WorkerConfiguration(kube_config=self.kube_config)
-        self.watcher_queue = multiprocessing.Queue()
+        self._manager = multiprocessing.Manager()
+        self.watcher_queue = self._manager.Queue()
```
Comment (ashb) [inline airflow/contrib/executors/kubernetes_executor.py:379]: Instead of using this here could we instead use a normal Queue and use get_nowait instead? Since the only use appears to be:

```python
        while not self.watcher_queue.empty():
            self.process_watcher_task()
```

could we instead do:

```python
        try:
            while True:
                task = self.watcher_queue.get_nowait()
                self.process_watcher_task(task)
        except queue.Empty:
            pass
```

(and change process_watcher_task fn to accept the event rather than having it pull it off the queue itself).

i.e. don't ever call `.empty()` on this queue.

File: airflow/contrib/executors/kubernetes_executor.py [hunk 4]
```diff
@@ -694,8 +698,8 @@ def start(self):
         # https://github.com/kubernetes-client/python/blob/master/kubernetes/docs
         # /CoreV1Api.md#list_namespaced_pod
         KubeResourceVersion.reset_resource_version()
-        self.task_queue = Queue()
-        self.result_queue = Queue()
+        self.task_queue = self._manager.Queue()
```
Comment (ashb) [inline airflow/contrib/executors/kubernetes_executor.py:712]: And here maybe we could use a multiprocessing.JoinableQueue instead.

File: tests/test_jobs.py (hunk 1 above)
Comment (potiuk) [reply tests/test_jobs.py]: Done.

File: airflow/executors/base_executor.py (hunk 2 above)
Comment (potiuk) [reply airflow/executors/base_executor.py]: 👍

File: airflow/contrib/executors/kubernetes_executor.py (hunk 3 above)
Comment (potiuk) [reply airflow/contrib/executors/kubernetes_executor.py:379]: I'd rather keep the current approach. In the implementation with managers, empty() has the added benefit that it will process all the events that were already put in the queue. With the get_nowait() there is a chance that some of those events already posted by other processes. Of course they will be picked up in the next heartbeat, but it somehow feels wrong. 

If you imagine that you have a busy scheduler, the get_nowait() will mean that there will pretty much always be some unprocessed entries left in the queue if you kill the scheduler. Not sure what heartbeat period usually is but it feels wrong to wait until next heartbeat. 

But I can be convinced otherwise (as I do not yet fully grasp all the details of this processing).

File: airflow/contrib/executors/kubernetes_executor.py (hunk 4 above)
Comment (potiuk) [reply airflow/contrib/executors/kubernetes_executor.py:712]: Agree. I think in this case we have no empty() problem.

Comment (potiuk) [reply airflow/contrib/executors/kubernetes_executor.py:712]: I also think we should get rid of the qsize() as it won't work on MacOS

File: airflow/contrib/executors/kubernetes_executor.py [hunk 5]
```diff
@@ -734,15 +741,17 @@ def sync(self):
 
         KubeResourceVersion.checkpoint_resource_version(last_resource_version)
 
-        for i in range(min((self.kube_config.worker_pods_creation_batch_size, self.task_queue.qsize()))):
-            task = self.task_queue.get()
-
+        for i in range(self.kube_config.worker_pods_creation_batch_size):
```
Comment (BasPH) [inline airflow/contrib/executors/kubernetes_executor.py]: Although the `for i in` was already in there, we don't use the `i` so no need to assign it:

```suggestion
        for _ in range(self.kube_config.worker_pods_creation_batch_size):
```

File: airflow/contrib/executors/kubernetes_executor.py [hunk 6]
```diff
@@ -734,15 +741,17 @@ def sync(self):
...
-            except ApiException:
-                self.log.exception('ApiException when attempting ' +
-                                   'to run task, re-queueing.')
-                self.task_queue.put(task)
+                task = self.task_queue.get_no_wait()
+                try:
+                    self.kube_scheduler.run_next(task)
+                except ApiException:
+                    self.log.exception('ApiException when attempting ' +
+                                       'to run task, re-queueing.')
```
Comment (BasPH) [inline airflow/contrib/executors/kubernetes_executor.py]: This fits on a single 110 char line

```python
                    self.log.exception('ApiException when attempting to run task, re-queueing.')
```

**Review by BasPH — COMMENTED** [review]
commented on 2 tiny nits

From a logic perspective, it appears okay to me although I'm not very experienced with the multiprocessing library. If the tests succeed and don't hang I'm okay with merging, to fix the current issues.

File: airflow/contrib/executors/kubernetes_executor.py (hunk 5 above)
Comment (potiuk) [reply airflow/contrib/executors/kubernetes_executor.py]: 👍

File: airflow/contrib/executors/kubernetes_executor.py (hunk 6 above)
Comment (potiuk) [reply airflow/contrib/executors/kubernetes_executor.py]: 👍

File: airflow/contrib/executors/kubernetes_executor.py (hunk 4 above)
Comment (potiuk) [reply airflow/contrib/executors/kubernetes_executor.py:712]: Ok. I took a deeper look and I discovered something else. Unlike the original multiprocessing.Queue(), the queue() returned by manager's Queue() method is actually a joinable queue already (both Python 2 and 3). And if I tried to use a direct (non-managed) Joinable Queue with get_nowait() I hit the same problem as the original LocalExecutor.test().\

At the end I am changing it to manager.Queue() and add task_done() to handle join separately.

BTW. I think task_done() was missing in the original implementation anyway as already .join() was used which could have caused unclean shutdown/hangs in the kubernetes executor. I kind of wonder how it worked before at all.

File: airflow/contrib/executors/kubernetes_executor.py (hunk 3 above)
Comment (potiuk) [reply airflow/contrib/executors/kubernetes_executor.py:379]: Ok. I convinced myself :). Once I use manager.Queue() everywhere it's actually good idea to use get_nowait here as it will be reliable as well.

Comment (potiuk) [reply airflow/contrib/executors/kubernetes_executor.py:379]: I replaced all the loops in kubernetes executor with the same pattern now, and included the task_done() calls as those are now Joinable queues.

Comment (potiuk) [reply airflow/contrib/executors/kubernetes_executor.py:379]: Just one small difference. I prefer the case where try/except is inside the while loop and we explicitly break . It's a bit clearer what our intention is rather than pass in outer exception.

File: airflow/contrib/executors/kubernetes_executor.py [hunk 7]
```diff
@@ -769,3 +793,7 @@ def _change_state(self, key, state, pod_id):
     def end(self):
         self.log.info('Shutting down Kubernetes executor')
         self.task_queue.join()
```
Comment (potiuk) [inline airflow/contrib/executors/kubernetes_executor.py]: I think here we had a problem here before because task_queue.join() was used but nowhere task_queue.task_done() was  executed. This could have resulted in hangs when shutting down Kubernetes Executor.

Comment (potiuk) [inline airflow/contrib/executors/kubernetes_executor.py:794]: I think we had problem here before. task_queue.join() was used but task_done() never called., so I think we could have some hangs when trying to stop gracefully KubernetesScheduler.

**Comment by potiuk** [comment]
@BasPH @Fokko @ashb -> I made some amendments now as I realised that Manager().Queue() is always a JoinableQueue (unlike multiprocessing.Queue). I updated the pattern everywhere to follow this general pattern below (also described it in the commit message). 

Since the managed queue is anyhow proxied to the shared one, get_nowait() will always return the  message if it already has been put to the queue by another process.

BTW. the get_nowait() message in case of managed queue is not really  "instant" as in case of standard multiprocessing.Queue - it will poll the shared process, but that's exactly why it is good for us :).

It also seems that in KubernetesExecutor there was a mistake. See the comment above (https://github.com/apache/airflow/pull/5200#discussion_r279615263) - if we've heard about inability to gracefully shutdown the KubernetesExecutor - that could be it.



```
while True:
   try:
       res = queue.get_nowait()
       try:
          .... do some processing
       finally:
           queue.task_done()
   except Empty:
       break

```

File: airflow/contrib/executors/kubernetes_executor.py [hunk 8]
```diff
@@ -440,11 +443,18 @@ def sync(self):
...
-    def process_watcher_task(self):
-        pod_id, state, labels, resource_version = self.watcher_queue.get()
+        while True:
+            try:
+                task = self.watcher_queue.get_nowait()
+                try:
+                    self.process_watcher_task(task)
+                finally:
+                    self.watcher_queue.task_done()
+            except queue.Empty:
```
Comment (milton0825) [inline airflow/contrib/executors/kubernetes_executor.py]:
```suggestion
            except Empty:
```

**Review by milton0825 — APPROVED** [review]
Thanks for the fix. LGTM.

**Comment by potiuk** [comment]
One last build and I think we are good to go :)

**Comment by potiuk** [comment]
Looks like we are good to go. Last chance to take a look @BasPH @ashb @Fokko :)

**Comment by ashb** [comment]
That `.join()` issue could be the cause of num_runs not working for Kube executor - https://issues.apache.org/jira/browse/AIRFLOW-4424.
